cargo-pgrx/src/command/upgrade.rs

```diff
@@ -0,0 +1,235 @@
+//LICENSE Portions Copyright 2019-2021 ZomboDB, LLC.
+//LICENSE
+//LICENSE Portions Copyright 2021-2023 Technology Concepts & Design, Inc.
+//LICENSE
+//LICENSE Portions Copyright 2023-2023 PgCentral Foundation, Inc. <contact@pgcentral.org>
+//LICENSE
+//LICENSE All rights reserved.
+//LICENSE
+//LICENSE Use of this source code is governed by the MIT license that can be found in the LICENSE file.
+use cargo_edit::{registry_url, Dependency};
+use eyre::eyre;
+use std::path::{Path, PathBuf};
+use toml_edit::KeyMut;
+use tracing::{debug, error, info, warn};
+
+use crate::CommandExecute;
+
+/// Upgrade pgrx crate versions in `Cargo.toml`.
+/// Defaults to latest.
+#[derive(clap::Args, Debug)]
+#[clap(author)]
+pub(crate) struct Upgrade {
+    /// Specify a version to upgrade to, rather than defaulting to the latest
+    /// version.
+    #[clap(long, short)]
+    pub(crate) target_version: Option<String>, //TODO: typed version not string
+
+    /// Path to the manifest file (usually Cargo.toml). Defaults to
+    /// "./Cargo.toml" in the working directory.
+    #[clap(long = "manifest-path", short)]
+    pub(crate) manifest_path: Option<PathBuf>,
+
+    // Flag to permit pre-release builds
+    #[clap(long, short)]
+    pub(crate) allow_prerelease: bool,
+}
+
+fn set_dep_source_version<A: AsRef<str>>(src: &mut cargo_edit::Source, new_version: A) {
+    match src {
+        cargo_edit::Source::Registry(reg) => reg.version = new_version.as_ref().to_string(),
+        cargo_edit::Source::Path(path) => path.version = Some(new_version.as_ref().to_string()),
+        cargo_edit::Source::Git(git) => git.version = Some(new_version.as_ref().to_string()),
+        cargo_edit::Source::Workspace(_ws) => {
+            error!(
+                "Cannot upgrade the version of \
+                a package because it is set in the \
+                workspace, please run `cargo pgrx \
+                upgrade` in the workspace directory."
+            );
+        }
+    }
+}
+fn get_dep_source_version(src: &cargo_edit::Source) -> Option<&String> {
+    match src {
+        cargo_edit::Source::Registry(reg) => Some(&reg.version),
```

Comment:
Should we ignore non-crates.io registries? Admittedly, I doubt it's common (but who knows what weird uses enterprise folks are getting up to with pgrx).

Reply:
If enterprise users are running their own registry and run `cargo pgrx upgrade`, they probably believe they have their own applicable versions of that crate (or their registry will fetch and cache the crates.io registry), so we should treat them equally.

Reply:
This assumption of course fails if enterprise software developers call the tool even when it is not applicable in their environment, and enterprise registry operators do not decide to make software that interoperates correctly (by usefully emulating crates.io or emitting useful errors). But otherwise, we should simply reward everyone for playing along. A custom registry *should* behave correctly.